sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -773,11 +770,18 @@ private void copyOne(String source, String destination) throws IOException {
     }
 
     @Override
-    public void remove(Collection<String> filenames) throws IOException {
-      for (String filename : filenames) {
+    public void removeDirectoryRecursively(String directory) throws IOException {
+      if (!new File(directory).exists()) {
+        LOG.debug("Directory {} already doesn't exist", directory);
+        return;
+      }
+      Collection<String> matches = factory.match(new File(directory, "*").getAbsolutePath());
```

Comment:
The method name indicates it deletes sub directories.
It requires sub directories becomes empty before they are removed.

I think FileIOChannelFactory.match() can guarantee that order by replacing preOrderTraversal with postOrderTraversal. (Long term, we probably want to specify the order with a MatchOptions)

For this PR, I suggest to choose a better method name. (We can enrich the functionality later when we need it)